### PR TITLE
activity/construction: fix ACT_MULTIPLE_CONSTRUCTION

### PR DESCRIPTION
**Pull request by andrei8l**
#### Summary
None

#### Purpose of change
Construction activity has a few problems:
1. it can get stuck in a loop fetching tools
2. it fails to find map items if the construction location is impassible (ex: `t_window`), or (in rare cases) if fetched items are just a little too far
3. it can't choose the appropriate construction variants from the same group like the construction UI (ex: barricaded window variants)
4. non-allied NPCs can work at the player's construction zones

Fixes: #57229

#### Describe the solution
See individual commit messages

#### Describe alternatives you've considered
~Just adding `BARRICADABLE_WINDOW` to `t_window_empty` to fix the freeze in the linked issue: this only masks one of the problems~ This is not even a solution. Adding that flag makes the window gain a sheet of glass for free.

#### Testing
Unit tests.

For #57229:
1. Load Yocemento save from the linked issue.
2. Approach the evac shelter. Game shouldn't freeze and the top window on the western side should be barricaded. After waiting a bit, there should be no "(...) is trying to find necessary items (...)" log spam

<details>
<summary>Additional Manual tests</summary>

3. Place a `t_window` terrain, then a "Board Up Window" construction zone on top of it
5. Start the "Construct Plots" activity and confirm the construction starts normally
6. Redo 3 but with `t_window_domestic`. Redo 4 and also confirm that byproducts (sheet, string, stick, nails) are spawned normally
7. Redo 5 but start the construction manually, then stop and start the activity construction. The construction should finish normally
</details>

#### Additional context
This entire chain desperately needs refactoring but this is the wrong time to do it so this patch will have to do for now.

### CODE REVIEW

**Comment by haveric** [comment]
It looks like you're removing one of the fixes I implemented a while ago and it's a little hard to tell if the changes you are making are accounting for it, so I just want to make sure you are aware of and testing for it: https://github.com/CleverRaven/Cataclysm-DDA/pull/55367

Either way, I'm glad somebody is reviewing this mess of code. Good luck!

**Comment by andrei8l** [comment]
I tested with "Build Dresser" (on top of items), "Deconstruct" (on top of a dresser with items), and "Mark firesource" (on top of a pile of items) and they all worked as expected. `keep_items` is a `post_flag` though so it can't block constructions (that's handled with `pre_special`).

**Comment by andrei8l** [comment]
Sorry for the many force pushes and status changes. I ran into several unexpected snags in this wonderful code chain. It should be good now, or your money back!

File: src/activity_item_handling.cpp [hunk 1]
```diff
@@ -1421,10 +1374,11 @@ static void add_basecamp_storage_to_loot_zone_list( zone_manager &mgr, const tri
     if( npc *const guy = dynamic_cast<npc *>( &you ) ) {
         map &here = get_map();
         if( guy->assigned_camp &&
-            mgr.has_near( zone_type_CAMP_STORAGE, here.getglobal( src_loc ), ACTIVITY_SEARCH_DISTANCE ) ) {
+            mgr.has_near( zone_type_CAMP_STORAGE, here.getglobal( src_loc ), ACTIVITY_SEARCH_DISTANCE ),
+            _fac_id( you ) ) {
```
Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp:1378]: It looks like parenthesis is positioned incorrectly in here. Should be:

image

**Pull request merged**